footil/log.py

```diff
@@ -39,3 +42,127 @@ def setup_logger(
     logging.basicConfig(
         format=LOG_FORMAT,
         level=log_level)
+
+
+def format_list_to_html(
+        line_height=1, collapse_cfg: Optional[dict] = None) -> str:
+    """Format list of strings into HTML string.
+
+    Lines are converted into HTML paragraphs. line-height attribute is
+    set for all paragraphs. This is default format.
+
+    Optionally can specify collapse_cfg to make part of text
+    "togglable" (aka read more/read less). Two ways are supported:
+        - bootrstrap collapse (default).
+        - attr_toggle on div that is "togglable".
+
+    In both cases max_lines key is required.
+
+    First case is used if max_lines is specified (maximum lines to
+    show initially), but not attr_toggle. Optionally collapse_id can be
+    passed to use it as anchor for bootsrap collapse toggle. Otherwise
+    str(uuid.uuid4()) value is used.
+
+    Second case is used if attr_toggle is used. This way only specified
+    custom attribute is added on div that wraps paragraphs that should
+    be hidden initially. Paragraphs toggle implementation must be done
+    externally in this case.
+
+    Args:
+        line_height: paragraph height (default: {1})
+        collapse_cfg: toggle show/hide part of text config
+            (default: {None}) Used keys:
+              - max_lines (int): number of paragraphs to show initially.
+              - collapse_id (str): anchor for collapse div
+                identification. If not set, will use randomly generated
+                ID. Bootstrap implementation only.
+              - attr_toggle (tuple): attribute used to toggle show/hide
+                part of text. Custom implementation only.
+
+    Returns:
+        HTML string
+    """
+    def build_lines(lines):
+        for line in lines:
+            # Assuming that line is one paragraph, so `\n` is not needed.
+            line = line.replace('\n', '')
+            with tag('p'):
+                text(line)
+
+    def build_boostrap_collapse(lines_to_hide):
+        # Default to random ID if none was specified. It has
+        # very low chance to run in collision, so there should
+        # be no problem.
+        collapse_id = (
+            collapse_cfg.get('collapse_id') or str(uuid.uuid4()))
+        with tag('div', id=collapse_id, klass='collapse'):
+            build_lines(lines_to_hide)
+        # Add button to toggle hidden lines.
+        with tag(
+            'a',
+            ('data-toggle', 'collapse'),
+            ('data-target', '#%s' % collapse_id),
+                klass='btn btn-link'):
+            text('Toggle More')
```

Comment:
Why not 'Show More'? :)

Reply:
Because it is single button. After you toggle show, it will show same button. If it would be 'Show More', it would not make sense, because it would actually mean show less:). Though this implementation will not be used for Odoo currently. Cause odoo has their own. At first I wanted to use my custom, but Odoo does strip away classes and custom attributes in tags. So to use it, I would need to hack odoo logic and force keep changes.

But because I already implemented this solution, I decided to keep it. There is second option, to specify custom attribute, which odoo uses itself to implement show more/show less.